src/api/unified/symbol_manager.hpp

```diff
@@ -105,10 +80,37 @@ class AFSymbolManager {
     std::shared_ptr<spdlog::logger> logger;
 };
 
-// Helper functions to ensure all the input arrays are on the active backend
-bool checkArray(af_backend activeBackend, const af_array a);
-bool checkArray(af_backend activeBackend, const af_array *a);
-bool checkArrays(af_backend activeBackend);
+namespace {
+bool checkArray(af_backend activeBackend, const af_array a) {
+    // Convert af_array into int to retrieve the backend info.
+    // See ArrayInfo.hpp for more
+    af_backend backend = (af_backend)0;
+
+    // This condition is required so that the invalid args tests for unified
+    // backend return the expected error rather than AF_ERR_ARR_BKND_MISMATCH
+    // Since a = 0, does not have a backend specified, it should be a
+    // AF_ERR_ARG instead of AF_ERR_ARR_BKND_MISMATCH
+    if (a == 0) return true;
+
+    af_get_backend_id(&backend, a);
+    return backend == activeBackend;
+}
+
+bool checkArray(af_backend activeBackend, const af_array* a) {
+    if (a) {
+        return checkArray(activeBackend, *a);
+    } else {
+        return true;
+    }
+}
+
+bool checkArrays(af_backend activeBackend) {
+    UNUSED(activeBackend);
+    // Dummy
+    return true;
+}
```

Comment:
Is there a specific reason you moved implementations from cpp to header ? This would later cause issue if and when this header is included in more than on translation unit.

Reply:
It's used in a template and unless if the implementation is in the header, the function will not be inlined.